gtsam/basis/Basis.h

```diff
@@ -0,0 +1,430 @@
+
+  for (int i = 0; i < x.cols(); i++) {
+    result.block(0, i * M, M, M).diagonal().array() = x(i);
+  }
+  return result;
+}
+
+/// CRTP Base class for function bases
+template <typename DERIVED>
+class GTSAM_EXPORT Basis {
+ public:
+  /**
+   * Call weights for all x in vector X.
+   * Returns M*N matrix where M is the size of the vector X,
+   * and N is the number of basis function.
+   */
+  static Matrix WeightMatrix(size_t N, const Vector& X) {
+    Matrix W(X.size(), N);
+    for (int i = 0; i < X.size(); i++)
+      W.row(i) = DERIVED::CalculateWeights(N, X(i));
+    return W;
+  }
+
+  /**
+   * Call weights for all x in vector X, with interval [a,b].
+   * Returns M*N matrix where M is the size of the vector X,
+   * and N is the number of basis function.
+   */
+  static Matrix WeightMatrix(size_t N, const Vector& X, double a, double b) {
+    Matrix W(X.size(), N);
+    for (int i = 0; i < X.size(); i++)
+      W.row(i) = DERIVED::CalculateWeights(N, X(i), a, b);
+    return W;
+  }
+
+  /**
+   * EvaluationFunctor calculate f(x;p) at a given `x`, applied to Parameters
+   * `p`. This functor is used to evaluate a parameterized function at a given
+   * scalar value x. When given a specific N*1 vector of Parameters, returns the
+   * scalar value of the function at x, possibly with Jacobian wrpt the
+   * parameters.
+   */
+  class EvaluationFunctor {
+   protected:
+    Weights weights_;
+
+   public:
+    /// For serialization
+    EvaluationFunctor() {}
+
+    /// Constructor with interval [a,b]
+    EvaluationFunctor(size_t N, double x)
+        : weights_(DERIVED::CalculateWeights(N, x)) {}
+
+    /// Constructor with interval [a,b]
+    EvaluationFunctor(size_t N, double x, double a, double b)
+        : weights_(DERIVED::CalculateWeights(N, x, a, b)) {}
+
+    /// Regular 1D evaluation
+    double apply(const typename DERIVED::Parameters& f,
+                 OptionalJacobian<-1, -1> H = boost::none) const {
+      if (H) *H = weights_;
+      return (weights_ * f)(0);
+    }
+
+    /// c++ sugar
+    double operator()(const typename DERIVED::Parameters& f,
+                      OptionalJacobian<-1, -1> H = boost::none) const {
+      return apply(f, H);  // might call apply in derived
+    }
+
+    void print(const std::string& s = "") const {
+      std::cout << s << (s != "" ? " " : "") << weights_ << std::endl;
+    }
+  };
+
+  /**
+   * VectorEvaluationFunctor at a given x, applied to ParameterMatrix<M>.
+   * This functor is used to evaluate a parameterized function at a given scalar
+   * value x. When given a specific M*N parameters, returns an M-vector the M
+   * corresponding functions at x, possibly with Jacobian wrpt the parameters.
+   */
+  template <int M>
+  class VectorEvaluationFunctor : protected EvaluationFunctor {
+   protected:
+    using VectorM = Eigen::Matrix<double, M, 1>;
+    using Jacobian = Eigen::Matrix<double, /*MxMN*/ M, -1>;
+    Jacobian H_;
+
+    void calculateJacobian() {
```

Comment:
Comment on what the Jacobian is, e.g.

```
/*
Calculate the `M*(M*N)` Jacobian of `apply` with respect to the M*N parameter matrix `P`.
We flatten assuming column-major order, e.g., if N=3 and M=2, we have
	H =[
		w(0) 0  w(1) 0  w(2) 0
		0 w(0)  0 w(1)  0 w(2)
	   ]
i.e., the Kronecker product of weights_ with the MxM identity matrix.
*/
```

Note *not* Doxygen as is private and not part of API